app/services/transaction_service/paypal_events.rb

```diff
@@ -50,34 +62,66 @@ def update_transaction_details(flow, details)
 
   # Privates
 
+  def set_session_context!(actor:, tx:)
+    existing_session_ctx = SessionContextStore.get
+
+    marketplace_session_ctx = {
+      marketplace_id: tx[:community_id],
+      marketplace_uuid: tx[:community_uuid]
+    }
+
+    user_session_ctx =
+      case actor
+      when :starter
+        {
+          user_id: tx[:starter_id],
+          user_uuid: tx[:starter_uuid]
+        }
+      when :author
+        {
+          user_id: tx[:listing_author_id],
+          user_uuid: tx[:listing_author_uuid]
+        }
+      when :unknown
+        # Unknown user
+        {}
+      else
+        raise ArgumentError.new("Unknown transition actor: #{actor}")
+      end
+
+    SessionContextStore.set(existing_session_ctx
+                              .merge(marketplace_session_ctx)
+                              .merge(user_session_ctx))
+  end
+
   def shipping_fields_present?(details)
     details.except(:status).values.any?
   end
 
   ## Mapping from payment transition to transaction transition
 
   TRANSITIONS = [
-    [:initiated_to_preauthorized,   [:initiated, :pending, :authorization]],
-    [:initiated_to_voided,          [:initiated, :voided]],
-    [:preauthorized_to_paid,        [:preauthorized, :completed]],
-    [:preauthorized_to_pending_ext, [:preauthorized, :pending]],
-    [:pending_ext_to_paid,          [:pending_ext, :completed]],
-    [:preauthorized_to_voided,      [:preauthorized, :voided]],
-    [:preauthorized_to_expired,     [:preauthorized, :expired]],
-    [:pending_ext_to_denied,        [:pending_ext, :denied]]
+    {name: :initiated_to_preauthorized,   actor: :starter, state: [:initiated, :pending, :authorization]},
+    {name: :initiated_to_voided,          actor: :starter, state: [:initiated, :voided]},
+    {name: :preauthorized_to_paid,        actor: :author,  state: [:preauthorized, :completed]},
+    {name: :preauthorized_to_pending_ext, actor: :author,  state: [:preauthorized, :pending]},
+    {name: :pending_ext_to_paid,          actor: :author,  state: [:pending_ext, :completed]},
+    {name: :preauthorized_to_voided,      actor: :author,  state: [:preauthorized, :voided]},
+    {name: :preauthorized_to_expired,     actor: :author,  state: [:preauthorized, :expired]},
+    {name: :pending_ext_to_denied,        actor: :author,  state: [:pending_ext, :denied]}
```

Comment:
I didn't realize setting the context here would be so involved. Anyway, this looks like a good way to do it.